File: examples/python/snippets/agents/workflow-agents/loop_agent_doc_improv_agent.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import asyncio
import os
from google.adk.agents import LoopAgent, LlmAgent, BaseAgent
from google.genai import types
# Use InMemoryRunner - no need for explicit SessionService
from google.adk.runners import InMemoryRunner
from google.adk.agents.invocation_context import InvocationContext
from typing import AsyncGenerator, Optional
from google.adk.events import Event, EventActions

# --- Constants ---
APP_NAME = "doc_writing_app"
USER_ID = "dev_user_01"
SESSION_ID = "loop_session_01"
GEMINI_MODEL = "gemini-2.0-flash"

# --- State Keys ---
# Using constants makes prompts and logic less error-prone
STATE_INITIAL_TOPIC = "initial_topic"
STATE_CURRENT_DOC = "current_document"
STATE_CRITICISM = "criticism"
STOP_WORD = "completed" # The exact word the CriticAgent should output to stop


# --8<-- [start:init]
# Part of agent.py --> Follow https://google.github.io/adk-docs/get-started/quickstart/ to learn the setup

# --- Agent Definitions ---

# Writer Agent
writer_agent = LlmAgent(
    name="WriterAgent",
    model=GEMINI_MODEL,
    # Improved Instruction: Use state key injection `{...}` and clear logic
    instruction=f"""You are a Creative Writing Assistant. Your goal is to iteratively write and refine a short document based on feedback.

IF the state key '{STATE_CURRENT_DOC}' does NOT exist or is empty:
Write a very short (1-2 sentence) story or document based on the topic provided in the state key '{STATE_INITIAL_TOPIC}'.

ELSE IF the state key '{STATE_CRITICISM}' exists and contains feedback (and is not '{STOP_WORD}'):
Refine the document currently in the state key '{STATE_CURRENT_DOC}' by thoughtfully applying the suggestions found in '{STATE_CRITICISM}'. Aim to improve the document based on the feedback.

Output *only* the new or refined document text. Do not add any introductory or concluding remarks.
""",
    description="Writes the initial document draft or refines it based on critique.",
    output_key=STATE_CURRENT_DOC # Saves output to state['current_document']
)

# Critic Agent
critic_agent = LlmAgent(
    name="CriticAgent",
    model=GEMINI_MODEL,
    # Improved Instruction: Specific output requirements for critique and stopping
    instruction=f"""You are a Constructive Critic AI. Your task is to review a document and provide feedback, or indicate completion.

Review the document provided in the session state key '{STATE_CURRENT_DOC}'.

IF the document can be improved:
Provide 1-2 brief, actionable suggestions for improvement (e.g., "Expand on the character's motivation.", "Add more sensory details."). Output *only* the critique text.

ELSE IF the document requires no further changes or is satisfactory:
Respond *exactly* with the word '{STOP_WORD}' and nothing else. This signals that the writing process is complete.

Do not add any explanations before or after your critique or the stop word.
""",
    description="Reviews the current document draft and provides critique or signals completion.",
    output_key=STATE_CRITICISM # Saves critique to state['criticism']
)

# Custom Agent to Check the Stop Condition
class CheckCondition(BaseAgent):
    """Checks the state for the critic's stop word and escalates if found."""
    async def _run_async_impl(self, ctx: InvocationContext) -> AsyncGenerator[Event, None]:
        # Retrieve the latest criticism from the state, default to "pending" if not set
        status = ctx.session.state.get(STATE_CRITICISM, "pending")
        # Normalize the status (lowercase, remove whitespace) for reliable comparison
        is_done = (status.lower().strip() == STOP_WORD)

        # Add print statement to show the condition check
        print(f"  [CheckCondition] Checking state '{STATE_CRITICISM}': Value='{status}', Stop condition met? {is_done}")

        # Escalate (stop the loop) if the condition is met
        yield Event(author=self.name, actions=EventActions(escalate=is_done))

# Create the LoopAgent
# It runs the sub_agents sequentially in each iteration until CheckCondition escalates or max_iterations is hit.
loop_agent = LoopAgent(
    name="IterativeWritingLoop",
    sub_agents=[
        writer_agent,    # Step 1: Write or refine
        critic_agent,    # Step 2: Critique or signal stop
        CheckCondition(name="CompletionChecker") # Step 3: Check if stopped
    ],
    max_iterations=5 # Limit loops to prevent infinite runs
)

# Assign the main loop agent as the root agent for the runner
root_agent = loop_agent

# --8<-- [end:init]

# Use InMemoryRunner
runner = InMemoryRunner(agent=root_agent, app_name=APP_NAME)
print(f"InMemoryRunner created for agent '{root_agent.name}'.")


# # Define an async interaction function with enhanced printing
async def call_loop_agent_async(initial_topic: str, user_id: str, session_id: str):
    """Runs the loop agent and prints intermediate outputs."""
    print(f"\n--- Starting Iterative Writing Process for topic: '{initial_topic}' ---")

    # Access the session service bundled within InMemoryRunner
    session_service = runner.session_service
    initial_state = {STATE_INITIAL_TOPIC: initial_topic}

    # Try to get the session first. If it doesn't exist, create it with the initial state.
    session = session_service.get_session(app_name=APP_NAME, user_id=user_id, session_id=session_id)
    if not session:
        print(f"  Session '{session_id}' not found, creating with initial state...")
        session = session_service.create_session(
            app_name=APP_NAME,
            user_id=user_id,
            session_id=session_id,
            state=initial_state
        )
        print(f"  Session '{session_id}' created.")
    else:
        print(f"  Session '{session_id}' exists. Ensuring initial topic is set.")
        try:
             stored_session = session_service.sessions[APP_NAME][user_id][session_id]
             stored_session.state[STATE_INITIAL_TOPIC] = initial_topic
        except KeyError:
             print(f"  Warning: Could not directly update stored session state for {session_id}. Relying on agent logic.")

    initial_message = types.Content(role='user', parts=[types.Part(text="Start writing process.")])
    iteration_count = 0
    loop_terminated_normally = False

    try:
        async for event in runner.run_async(
            user_id=user_id,
            session_id=session_id,
            new_message=initial_message,
        ):
            author_name = event.author or "System"
            is_final = event.is_final_response()
            print(f"  [Event] From: {author_name}, Final: {is_final}") # Basic event logging

            # Check if it's a final response from one of the main agents
            if is_final and event.content and event.content.parts:
                output_text = event.content.parts[0].text.strip()

                if author_name == writer_agent.name:
                    iteration_count += 1
                    print(f"\n[Iteration {iteration_count}] WriterAgent Output ({STATE_CURRENT_DOC}):")
                    print(output_text)

                elif author_name == critic_agent.name:
                    print(f"[Iteration {iteration_count}] CriticAgent Output ({STATE_CRITICISM}):")
                    print(output_text)
                    print(f"  (Saving to state key '{STATE_CRITICISM}')")

            # Check if the loop was terminated by the CheckCondition agent's escalation
            if event.actions and event.actions.escalate and author_name == "CompletionChecker":
                 print(f"\n--- Loop terminated by CompletionChecker (Condition Met: '{STOP_WORD}') ---")
                 loop_terminated_normally = True
                 # Runner handles escalation, loop continues until exhausted or next natural break

            # --- Corrected Error Check ---
            # Check if the event contains an error message instead of using is_error()
            elif event.error_message:
                 print(f"  -> Error from {author_name}: {event.error_message}")

    except Exception as e:
        print(f"\n❌ An error occurred during agent execution: {e}")

    if not loop_terminated_normally:
         print(f"\n--- Loop finished (Max iterations {loop_agent.max_iterations} reached or other termination) ---")

    # Optional: Inspect final state
    final_state = runner.session_service.get_session(app_name=APP_NAME,user_id=user_id, session_id=session_id)
    print("\n--- Final Session State ---")
    print(final_state if final_state else "State not found or empty.")
    print("-" * 30)

# # Define main async function to run the interaction
# async def main():
#     """Main function to execute the agent interaction."""
topic = "a journey to a newly discovered planet"
# topic = "the first day of training for a clumsy dragon rider"
# topic = "the challenges of communicating with a plant-based alien species"
# topic = "life inside a giant library where books are portals"
await call_loop_agent_async(topic, user_id=USER_ID, session_id=SESSION_ID)
# asyncio.run(call_loop_agent_async(topic, user_id=USER_ID, session_id=SESSION_ID))

File: examples/typescript/snippets/agents/workflow-agents/loop_agent_doc_improv_agent.ts
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// --8<-- [start:init]
// Part of agent.ts --> Follow https://google.github.io/adk-docs/get-started/quickstart/ to learn the setup

import { LoopAgent, LlmAgent, SequentialAgent, InMemoryRunner, FunctionTool, ToolContext } from '@google/adk';
import { Content } from '@google/genai';
import { z } from 'zod';

// --- Constants ---
const APP_NAME = "doc_writing_app_v3"; // New App Name
const USER_ID = "dev_user_01";
const SESSION_ID_BASE = "loop_exit_tool_session"; // New Base Session ID
const GEMINI_MODEL = "gemini-2.0-flash";
const STATE_INITIAL_TOPIC = "initial_topic";

// --- State Keys ---
const STATE_CURRENT_DOC = "current_document";
const STATE_CRITICISM = "criticism";
// Define the exact phrase the Critic should use to signal completion
const COMPLETION_PHRASE = "No major issues found.";

// --- Tool Definition ---
function exitLoop(toolContext: ToolContext): Record<string, unknown> {
  /** Call this function ONLY when the critique indicates no further changes are needed, signaling the iterative process should end. */
  console.log(`  [Tool Call] exit_loop triggered by ${toolContext.agentName}`);
  toolContext.actions.escalate = true;
  // Return empty object as tools should typically return JSON-serializable output
  return {};
}

const exitLoopTool = new FunctionTool({
    name: 'exit_loop',
    description: 'Call this function ONLY when the critique indicates no further changes are needed, signaling the iterative process should end.',
    parameters: z.object({}),
    execute: exitLoop,
});

// --- Agent Definitions ---

// STEP 1: Initial Writer Agent (Runs ONCE at the beginning)
const initialWriterAgent = new LlmAgent({
    name: "InitialWriterAgent",
    model: GEMINI_MODEL,
    includeContents: 'none',
    // MODIFIED Instruction: Ask for a slightly more developed start
    instruction: `You are a Creative Writing Assistant tasked with starting a story.
    Write the *first draft* of a short story (aim for 2-4 sentences).
    Base the content *only* on the topic provided below. Try to introduce a specific element (like a character, a setting detail, or a starting action) to make it engaging.
    Topic: {{initial_topic}}

    Output *only* the story/document text. Do not add introductions or explanations.
`,
    description: "Writes the initial document draft based on the topic, aiming for some initial substance.",
    outputKey: STATE_CURRENT_DOC
});

// STEP 2a: Critic Agent (Inside the Refinement Loop)
const criticAgentInLoop = new LlmAgent({
    name: "CriticAgent",
    model: GEMINI_MODEL,
    includeContents: 'none',
    // MODIFIED Instruction: More nuanced completion criteria, look for clear improvement paths.
    instruction: `You are a Constructive Critic AI reviewing a short document draft (typically 2-6 sentences). Your goal is balanced feedback.

    **Document to Review:**
    ```
    {{current_document}}
    ```

    **Task:**
    Review the document for clarity, engagement, and basic coherence according to the initial topic (if known).

    IF you identify 1-2 *clear and actionable* ways the document could be improved to better capture the topic or enhance reader engagement (e.g., "Needs a stronger opening sentence", "Clarify the character's goal"):
    Provide these specific suggestions concisely. Output *only* the critique text.

    ELSE IF the document is coherent, addresses the topic adequately for its length, and has no glaring errors or obvious omissions:
    Respond *exactly* with the phrase "${COMPLETION_PHRASE}" and nothing else. It doesn't need to be perfect, just functionally complete for this stage. Avoid suggesting purely subjective stylistic preferences if the core is sound.

    Do not add explanations. Output only the critique OR the exact completion phrase.
`,
    description: "Reviews the current draft, providing critique if clear improvements are needed, otherwise signals completion.",
    outputKey: STATE_CRITICISM
});


// STEP 2b: Refiner/Exiter Agent (Inside the Refinement Loop)
const refinerAgentInLoop = new LlmAgent({
    name: "RefinerAgent",
    model: GEMINI_MODEL,
    // Relies solely on state via placeholders
    includeContents: 'none',
    instruction: `You are a Creative Writing Assistant refining a document based on feedback OR exiting the process.
    **Current Document:**
    ```
    {{current_document}}
    ```
    **Critique/Suggestions:**
    {{criticism}}

    **Task:**
    Analyze the 'Critique/Suggestions'.
    IF the critique is *exactly* "${COMPLETION_PHRASE}":
    You MUST call the 'exit_loop' function. Do not output any text.
    ELSE (the critique contains actionable feedback):
    Carefully apply the suggestions to improve the 'Current Document'. Output *only* the refined document text.

    Do not add explanations. Either output the refined document OR call the exit_loop function.
`,
    description: "Refines the document based on critique, or calls exit_loop if critique indicates completion.",
    tools: [exitLoopTool], // Provide the exit_loop tool
    outputKey: STATE_CURRENT_DOC // Overwrites state['current_document'] with the refined version
});


// STEP 2: Refinement Loop Agent
const refinementLoop = new LoopAgent({
    name: "RefinementLoop",
    // Agent order is crucial: Critique first, then Refine/Exit
    subAgents: [
        criticAgentInLoop,
        refinerAgentInLoop,
    ],
    maxIterations: 5 // Limit loops
});

// STEP 3: Overall Sequential Pipeline
// For ADK tools compatibility, the root agent must be named `root_agent`
const rootAgent = new SequentialAgent({
    name: "IterativeWritingPipeline",
    subAgents: [
        initialWriterAgent, // Run first to create initial doc
        refinementLoop       // Then run the critique/refine loop
    ],
    description: "Writes an initial document and then iteratively refines it with critique using an exit tool."
});
// --8<-- [end:init]


// --- Running the Agent on Notebooks/Scripts ---
// const runner = new InMemoryRunner({agent: rootAgent, appName: APP_NAME});
// console.log(`InMemoryRunner created for agent '${rootAgent.name}'.`);


// // Interaction function (Modified to show agent names and flow)
// async function callPipelineAsync(initialTopic: string, userId: string, sessionId: string) {
//     console.log(`
--- Starting Iterative Writing Pipeline (Exit Tool) for topic: '${initialTopic}' ---`);
//     const sessionService = runner.sessionService;
//     const initialState = {[STATE_INITIAL_TOPIC]: initialTopic};
//     // Explicitly create/check session BEFORE run
//     let session = await sessionService.getSession({appName: APP_NAME, userId, sessionId});
//     if (!session) {
//         console.log(`  Session '${sessionId}' not found, creating with initial state...`);
//         session = await sessionService.createSession({appName: APP_NAME, userId, sessionId, state: initialState});
//         console.log(`  Session '${sessionId}' created.`);
//     } else {
//         console.log(`  Session '${sessionId}' exists. Resetting state for new run.`);
//         // In a real app, you might have a more robust way to update state
//         session.state = initialState;
//         await sessionService.updateSession(session);
//     }

//     const initialMessage: Content = {role: 'user', parts: [{text: "Start the writing pipeline."}]};
//     let loopIteration = 0;
//     let pipelineFinishedViaExit = false;
//     let lastKnownDoc = "No document generated."; // Store the last document output

//     try {
//         for await (const event of runner.run({userId, sessionId, newMessage: initialMessage})) {
//             const authorName = event.author || "System";
//             const isFinal = event.isFinalResponse;
//             console.log(`  [Event] From: ${authorName}, Final: ${isFinal}`);

//             // Display output from each main agent when it finishes
//             if (isFinal && event.content && event.content.parts) {
//                 const outputText = event.content.parts[0].text!.trim();

//                 if (authorName === initialWriterAgent.name) {
//                     console.log(`
[Initial Draft] By ${authorName} (${STATE_CURRENT_DOC}):`);
//                     console.log(outputText);
//                     lastKnownDoc = outputText;
//                 } else if (authorName === criticAgentInLoop.name) {
//                     loopIteration++;
//                     console.log(`
[Loop Iteration ${loopIteration}] Critique by ${authorName} (${STATE_CRITICISM}):`);
//                     console.log(outputText);
//                     console.log(`  (Saving to state key '${STATE_CRITICISM}')`);
//                 } else if (authorName === refinerAgentInLoop.name) {
//                     // Only print if it actually refined (didn't call exit_loop)
//                     if (!event.actions?.escalate) { // Check if exit wasn't triggered in *this* event's actions
//                         console.log(`[Loop Iteration ${loopIteration}] Refinement by ${authorName} (${STATE_CURRENT_DOC}):`);
//                         console.log(outputText);
//                         lastKnownDoc = outputText;
//                         console.log(`  (Overwriting state key '${STATE_CURRENT_DOC}')`);
//                     }
//                 }
//             }

//             if (event.actions?.escalate) {
//                  console.log(`
--- Refinement Loop terminated (Escalation detected) ---`);
//                  pipelineFinishedViaExit = true;
//                  break;
//             } else if (event.errorMessage) {
//                  console.log(`  -> Error from ${authorName}: ${event.errorMessage}`);
//                  break; // Stop on error
//             }
//         }
//     } catch (e) {
//         console.log(`
❌ An error occurred during agent execution: ${e}`);
//     }

//     if (pipelineFinishedViaExit) {
//         console.log(`
--- Pipeline Finished (Terminated by exit_loop) ---`);
//     } else {
//         console.log(`
--- Pipeline Finished (Max iterations ${refinementLoop.maxIterations} reached or error) ---`);
//     }

//     console.log(`Final Document Output:
${lastKnownDoc}`);

//     // Final state retrieval
//     const finalSessionObject = await runner.sessionService.getSession({appName: APP_NAME, userId, sessionId});
//     console.log("\n--- Final Session State ---");
//     if (finalSessionObject) {
//         console.log(finalSessionObject.state);
//     } else {
//         console.log("State not found (Final session object could not be retrieved).");
//     }
//     console.log("-".repeat(30));
// }


// const topic = "a robot developing unexpected emotions";
// // const topic = "the challenges of communicating with a plant-based alien species";

// // A simple hash function for demonstration
// const simpleHash = (s: string) => s.split('').reduce((a,b)=>{a=((a<<5)-a)+b.charCodeAt(0);return a&a},0);
// const sessionId = `${SESSION_ID_BASE}_${Math.abs(simpleHash(topic)) % 1000}`; // Unique session ID

// await callPipelineAsync(topic, USER_ID, sessionId);
